data_structures/tree_234.cpp

```diff
@@ -0,0 +1,1085 @@
+     * requirements are not fulfilled.
+     */
+    bool TryRightRotate(Node *parent, Node *to_child);
+
+    /**
+     * @brief do the actual right rotate operation
+     *
+     * Given parent node, and the pivot item index, the right rotate operation
+     * is uniquely identified. The function assume the requirements are
+     * fulfilled and won't do any extra check. This function is call by
+     * TryRightRotate(), and the condition checking should be done before call
+     * it.
+     *
+     * @param parent the parent node in this right rotate operation
+     * @param index the pivot item index of this right rotate operation.
+     */
+    void RightRotate(Node *parent, int index);
+
+    /**
+     * @brief do the actual left rotate operation
+     *
+     * Given parent node, and the pivot item index, the left rotate operation is
+     * uniquely identified. The function assume the requirements are fulfilled
+     * and won't do any extra check. This function is call by TryLeftRotate(),
+     * and the condition checking should be done before call it.
+     *
+     * @param parent the parent node in this right rotate operation
+     * @param index the pivot item index of this right rotate operation.
+     */
+    void LeftRotate(Node *parent, int index);
+
+    /**
+     * @brief main function implement the pre-merge remove operation
+     * @param node the tree to remove item from
+     * @param item item to remove
+     * @return return true if remove success, false otherwise
+     * */
+    bool RemovePreMerge(Node *node, int item);
+
+    /**
+     * @brief merge the item at index of the parent node, and its left and right
+     * child
+     *
+     * the left and right child node must be 2-node. The 3 items will be merged
+     * into a 4-node. In our case the parent can be a 2-node iff it is the root.
+     * Otherwise, it must be 3-node or 4-node.
+     *
+     * @param parent the parent node in the merging operation
+     * @param index the item index of the parent node that involved in the
+     * merging
+     * @return return the merged 4-node
+     * */
+    Node *Merge(Node *parent, int index);
+
+    /** @brief recursive release the tree */
+    void DeleteNode(Node *tree);
+
+    /** @brief In-order traverse the tree, print items */
+    void Traverse(Node *tree);
+
+    /** @brief Print the tree to a dot file. You can convert it to picture with
+     * graphviz*/
+    void PrintNode(FILE *f, Node *node, int parent_index, int index,
+                   int parent_ptr_index);
+
+    /** @brief Print the tree with all its empty item also printed with a '-'.
+     * This function is an alternative to PrintNode() */
+    void PrintNode2(FILE *f, Node *node, int parent_index, int index,
+                    int parent_ptr_index);
+
+    Node *root_{nullptr};  ///< root node of the tree
+};
+
+Tree234::~Tree234() { DeleteNode(root_); }
+
+void Tree234::DeleteNode(Node *tree) {
```

Comment:
Missing function documentation. Even if the prototype is documented, it's also good to document it here (same in other parts of the code).

Reply:
Do you mean copy the document here again?